jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -188,42 +269,42 @@
                 return this;
             }
 
+            /**
+             * Configure the index to be mutable.
+             *
+             * @return a {@link Mutable.Builder} configured like this builder.
+             */
+            public Mutable.Builder<V> mutable()
+            {
+                return this;
+            }
+
             /**
              * Build a {@link Mutable} instance.
              * @return a {@link Mutable} instance.
              */
             public Mutable<V> build()
             {
-                if (contents != null && maxCapacity == 0)
-                    throw new IllegalStateException("Cannot create a mutable index with maxCapacity=0 and some contents");
-
-                // TODO we need to consider large size and alphabet when picking a trie impl
-                Mutable<V> result;
-                if (maxCapacity > 0)
-                {
-                    result = new ArrayTernaryTrie<>(!caseSensitive, maxCapacity);
-                }
-                else if (maxCapacity < 0)
-                {
-                    if (caseSensitive)
-                        result = new ArrayTernaryTrie.Growing<>(false, 512, 512);
-                    else
-                        result = new TreeTrie<>();
-                }
-                else
-                {
-                    result = EmptyTrie.instance(caseSensitive);
-                }
-
-                if (contents != null)
-                {
-                    for (Map.Entry<String, V> entry : contents.entrySet())
-                    {
-                        if (!result.put(entry.getKey(), entry.getValue()))
-                            throw new AssertionError("Index capacity exceeded at " + entry.getKey());
-                    }
-                }
-                return result;
+                if (maxCapacity == 0)
+                    return EmptyTrie.instance(caseSensitive);
+
+                // Work out needed capacity
+                int capacity = (contents == null) ? 0 : AbstractTrie.requiredCapacity(contents.keySet(), caseSensitive);
+
+                // check capacities
+                if (maxCapacity >= 0 && capacity > maxCapacity)
+                    throw new IllegalStateException("Insufficient maxCapacity for contents");
```

Comment:
Is `AbstractTrie.requiredCapacity` reporting the exact amount of required bytes for each and every trie implementation? Otherwise we have a problem.

Reply:
It is if we assume all our Trie implementations can handle a key of "".
The other thing to do is to just not handle a "" and get rid of the whole +1 logic.... but then the nature of these beasts is that they have a root and the root can store a value, so hmmmm  keeping the +1 is probably simpler.    I'll document.

Reply:
`ArrayTernaryTrie` needs more than just +1, see https://github.com/eclipse/jetty.project/blob/2662be0a05e14fad80defd722bc191ba201c5e8b/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTernaryTrie.java#L125

Reply:
That comment is wrong, the +1 is enough.  Anyway, that class is deprecated and never directly called from the Index factory now.  I'll fix the comment.

Reply:
more javadoc improvements

Reply:
The `ArrayTernaryTrie` capacity calculation is really a pain, but since that implementation isn't used by the `Index` builders anymore, this looks good enough.